src/stock.rs

```diff
@@ -123,19 +122,47 @@ impl StockMessage {
 }
 
 impl Context {
+    /// Set the stock string for the [StockMessage].
+    ///
+    pub fn set_stock_translation(
+        &self,
+        id: StockMessage,
+        stockstring: String,
+    ) -> Result<(), Error> {
+        if stockstring.contains("%1") && !id.fallback().contains("%1") {
+            bail!(
+                "translation {} contains invalid %1 placeholder, default is {}",
+                stockstring,
+                id.fallback()
+            );
+        }
+        if stockstring.contains("%2") && !id.fallback().contains("%2") {
+            bail!(
+                "translation {} contains invalid %2 placeholder, default is {}",
+                stockstring,
+                id.fallback()
+            );
+        }
+        self.translated_stockstrings
+            .write()
+            .unwrap()
+            .insert(id as usize, stockstring);
+        Ok(())
+    }
+
     /// Return the stock string for the [StockMessage].
     ///
-    /// If the context callback responds with a string to use, e.g. a
-    /// translation, then this string will be returned.  Otherwise a
-    /// default (English) string is returned.
+    /// Return a translation (if it was set with set_stock_translation before)
+    /// or a default (English) string.
     pub fn stock_str(&self, id: StockMessage) -> Cow<str> {
-        let ptr = self.call_cb(Event::GetString { id, count: 0 }) as *mut libc::c_char;
-        if ptr.is_null() {
-            Cow::Borrowed(id.fallback())
-        } else {
-            let ret = to_string_lossy(ptr);
-            unsafe { libc::free(ptr as *mut libc::c_void) };
-            Cow::Owned(ret)
+        match self
+            .translated_stockstrings
+            .read()
+            .unwrap()
+            .get(&(id as usize))
+        {
+            Some(ref x) => Cow::Owned(x.to_string()),
+            None => Cow::Borrowed(id.fallback()),
```

Comment:
Does this now still need to return a `Cow<str>` or could it just return a `&str`?  What are the guarantees that the strings stay in the hashmap while the result of this call is used?

Reply:
i think the hashmap can be modified at any time. eg. if the system-language is changed, the ui may replace all translations in the hashmap by calling dc_set_stock_translation() again for each string.

Reply:
thought about this too, but didn't want to change all calling places now.  right now there is nothing preventing arbitrary set_stock_translation calls at arbtirary times.